ODIN_II/SRC/HardSoftLogicMixer.cpp

```diff
@@ -0,0 +1,212 @@
+        switch (i) {
+            case mix_hard_blocks::MULTIPLIERS:
+                if (_enabledOptimizations[mix_hard_blocks::MULTIPLIERS]) {
+                    choose_hard_blocks(netlist, mix_hard_blocks::MULTIPLIERS);
+                }
+                break;
+            default:
+                error_message(INC_IMPLEMENTATION, unknown_location, "%s",
+                              "Optimization with number: %i does not have an implementation inside of HardSoftLogicMixer::map_deferred_blocks\n", i);
+                break;
+        }
+    }
+    soft_map_remaining_nodes(netlist);
+}
+
+void HardSoftLogicMixer::grid_statistics() {
+    if (true == _allOptsDisabled) {
+        return;
+    }
+
+    for (int opt = 0; opt < mix_hard_blocks::Count; opt++) {
+        if (true == _enabledOptimizations[opt]) {
+            // with development for fixed_layout, this value will change
+            int availableHardBlocks = INT_MAX;
+            int hardBlocksNeeded = hard_blocks_needed(opt);
+            int hardBlocksCount = availableHardBlocks;
+
+            if (hardBlocksCount > hardBlocksNeeded) {
+                hardBlocksCount = hardBlocksNeeded;
+            }
+
+            if (hardBlocksCount < _hardBlocksCount[opt]) {
+                _hardBlocksCount[opt] = hardBlocksCount;
+            }
+        }
+    }
+    scale_counts();
+}
+
+void HardSoftLogicMixer::scale_counts() {
+    for (int i = 0; i < mix_hard_blocks::Count; i++) {
+        if (_hardBlocksMixingRatio[i] != -1) {
+            _hardBlocksCount[i] = _hardBlocksCount[i] * _hardBlocksMixingRatio[i];
+        }
+    }
+}
+
+int HardSoftLogicMixer::hard_blocks_needed(int opt) {
+    return _candidate_nodes[opt].size();
+}
+
+void HardSoftLogicMixer::soft_map_remaining_nodes(netlist_t* netlist) {
+    for (unsigned int i = 0; i < mix_hard_blocks::Count; i++) {
+        for (unsigned int j = 0; j < _candidate_nodes[i].size(); j++) {
+            switch (i) {
+                case mix_hard_blocks::MULTIPLIERS:
+                    instantiate_simple_soft_multiplier(_candidate_nodes[i][j], PARTIAL_MAP_TRAVERSE_VALUE, netlist);
+                    break;
+                default:
+                    error_message(INC_IMPLEMENTATION, unknown_location, "%s",
+                                  "Cleanup for optimization with number: Optimization with number: %i does not have an implementation inside of HardSoftLogicMixer::soft_map_remaining_nodes\n", i);
+                    break;
+            }
+        }
+    }
+}
+
+void HardSoftLogicMixer::choose_hard_blocks(netlist_t* netlist, mix_hard_blocks type) {
+    if (type == mix_hard_blocks::Count) {
+        error_message(INC_IMPLEMENTATION, unknown_location, "%s",
+                      "Running choseHardBlocks with Count should never happen");
+        exit(6);
+    }
+    std::vector<nnode_t*>& nodesVector = _candidate_nodes[type];
+    int nodesCount = nodesVector.size();
+    int* costs = new int[nodesCount];
```

Comment:
just use a std::vector

Reply:
Thank you for the answer,
Do you mean instead of 
```std::vector<nnode_t*>& nodesVector```
or 
```int* costs  ?```

Reply:
After thinking more about it, I think the cost is something that is highly relevant for a lot of optimization, you could tack it on the stats_t in the nnode_t or directly in the nnode_t. 
I think it is worth it to keep it in there for future usage.

```
// in odin_types.h

struct nnode_t {
    ...
    /* by default a node has a cost of one ?? or something better? */
    int cost = 1;
   ...
};
```

so that you don't need the 
```
int* costs = new int[nodesCount]
```

just make sure to make the name a bit more explicit... maybe optimization_cost, or path_cost

Reply:
@jeanlego I did some changes before I got your answer, how does the proposed look like?